test/unit/ipam_test.rb

```diff
@@ -0,0 +1,90 @@
+require 'test_helper'
+
+class IPAMTest < ActiveSupport::TestCase
+  test 'should raise an exception if given mac is invalid' do
+    s = FactoryGirl.build(:subnet_ipv6, :network => '2001:db8::')
+    exception = assert_raise Foreman::Exception do
+      IPAM::Eui64.new(:subnet => s, :mac => 'invalid')
+    end
+    assert_includes exception.message, 'not a valid MAC address'
+  end
+
+  context 'dhcp' do
+    test "should find unused IP on proxy if proxy is set" do
+      subnet = FactoryGirl.create(:subnet_ipv4, :ipam_dhcp, :name => 'my_subnet', :network => '192.168.1.0')
+      subnet.stubs(:dhcp? => true)
+      subnet.stubs(:dhcp => mock('attribute', :url => 'proxy.example.com'))
+      fake_proxy = mock("dhcp_proxy")
+      fake_proxy.stubs(:unused_ip => {'ip' => '192.168.1.25'})
+      subnet.stubs(:dhcp_proxy => fake_proxy)
+      ipam = IPAM::Dhcp.new(:subnet => subnet, :mac => '00:11:22:33:44:55')
+      assert_equal '192.168.1.25', ipam.suggest_ip
+    end
+  end
+
+  context 'internal db' do
+    test "should find unused IP in internal DB if proxy is not set" do
+      host = FactoryGirl.create(:host)
+      subnet = FactoryGirl.create(:subnet_ipv4, :name => 'my_subnet', :network => '192.168.2.0',
+                                  :ipam => IPAM::MODES[:db])
+      ipam = IPAM::Db.new(:subnet => subnet)
+      assert_equal '192.168.2.1', ipam.suggest_ip
+
+      subnet.reload
+      FactoryGirl.create(:nic_managed, :ip => '192.168.2.1', :subnet_id => subnet.id, :host => host, :mac => '00:00:00:00:00:01')
+      FactoryGirl.create(:nic_managed, :ip => '192.168.2.2', :subnet_id => subnet.id, :host => host, :mac => '00:00:00:00:00:02')
+      assert_equal '192.168.2.3', ipam.suggest_ip
+    end
+
+    test "should find unused IP excluding named values in internal DB if proxy is not set" do
+      host = FactoryGirl.create(:host)
+      subnet = FactoryGirl.create(:subnet_ipv4, :name => 'my_subnet', :network => '192.168.2.0',
+                                  :ipam => IPAM::MODES[:db])
+      ipam = IPAM::Db.new(:subnet => subnet, :excluded_ips => ['192.168.2.1', '192.168.2.2'])
+      assert_equal '192.168.2.3', ipam.suggest_ip
+
+      subnet.reload
+      FactoryGirl.create(:nic_managed, :ip => '192.168.2.1', :subnet_id => subnet.id, :host => host, :mac => '00:00:00:00:00:01')
+      FactoryGirl.create(:nic_managed, :ip => '192.168.2.2', :subnet_id => subnet.id, :host => host, :mac => '00:00:00:00:00:02')
+      ipam = IPAM::Db.new(:subnet => subnet, :excluded_ips => ['192.168.2.3'])
+      assert_equal '192.168.2.4', ipam.suggest_ip
+    end
+
+    test "should respect subnet from and to if it's set" do
+      host = FactoryGirl.create(:host)
```

Comment:
Ditto, no need to create hosts and nic_manageds https://github.com/theforeman/foreman/pull/3442/files#r61550858